### ISSUE TITLE
Should not use a hardcoded shell; fails on Alpine systems.

### ISSUE THREAD

**Issue by Skrylar** [pyinvoke/invoke#752]
pyinvoke should not be hard-coded to want to use `/bin/bash`. This is not appropriate Unix behavior.

On an Alpine system you will receive this error: `FileNotFoundError: [Errno 2] No such file or directory: '/bin/bash'`

There are workarounds for this (installing bash, symlinking /bin/sh, ...) but I believe it is customary to use whatever is in $SHELL or the login shell.

**Comment by miso-belica** [comment]
@Skrylar I believe this is customizable. You should be able to set any shell by [config or ENV variable](http://docs.pyinvoke.org/en/stable/concepts/configuration.html). Also, you can pass named argument `shell='/bin/sh'` or whatever you want to the [method `run`](http://docs.pyinvoke.org/en/stable/api/runners.html#invoke.runners.Runner.run) to overwrite it.

```yaml
run:
  shell: C:\Windows\System32\cmd.exe
```

**Comment by miso-belica** [comment]
@Skrylar yep, that is what I wrote. Invoke is not using `$SHELL` there are different ENV variables but the shell is not hardcoded - you can change it. If you click on the links you see that there are ENV variables to use or config file to overwrite the shell.

I agree, there could be a much better message though. This is unfortunate. But you don't need to wait for this, you can set any shell you like even now :)

**Comment by sprat** [comment]
Using $SHELL is not a good idea IMHO since the local machine may be very different from remote ones.

Anyway, why not use /bin/sh by default which is available everywhere AFAIK?

**Comment by talvdav** [comment]
> Using $SHELL is not a good idea IMHO since the local machine may be very different from remote ones. 

using /bin/bash is also not a good idea either

In the Config class it is checked to see if on windows and change the shell acording the result of os.environ.get, if not shell is set to /bin/bash.

I think it would be better to first check os.environ.get('SHELL') and use this and if it fails default do '/bin/sh'

**Comment by Varriount** [comment]
@talvdav I don't believe using *any* shell by default is a good idea - it's unpredictable, not cross-platform, and prone to vulunerabilities (`"rm -rf"` anyone?). I was quite surprised when I found out that pyinvoke defaults to using one.

**Comment by aastashov** [comment]
Faced with the same problem. It works for me:

```python
import os
from invoke import Context, Result, task

@task
def some_task(context: Context) -> None:
    if not os.path.exists("/bin/bash"):
        context.config.run.shell = "/bin/sh"

    git_status: Result = context.run("git status")
    print(git_status.stdout)
```

Link to the authors' suggestion to use the config:
https://github.com/pyinvoke/invoke/blob/2.0.0/invoke/config.py#L435

**Comment by robertschweizer** [comment]
I just hit this as well. A nice workaround for an entire `tasks.py` file is adding this at the top of the file:

```py
import os

os.environ["INVOKE_RUN_SHELL"] = "/bin/sh"
```
